Composer/plugins/azurePublish/src/node/azureProvisionUtil.ts

```diff
@@ -0,0 +1,323 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { SubscriptionClient } from '@azure/arm-subscriptions';
+import { Subscription } from '@azure/arm-subscriptions/esm/models';
+import { ResourceManagementClient } from '@azure/arm-resources';
+import { ResourceGroup, GenericResource } from '@azure/arm-resources/esm/models';
+import { AzureBotService } from '@azure/arm-botservice';
+import { WebSiteManagementClient } from '@azure/arm-appservice';
+import { ResourceNameAvailability } from '@azure/arm-appservice/esm/models';
+import { CheckNameAvailabilityResponseBody } from '@azure/arm-botservice/esm/models';
+
+export enum AzureAPIStatus {
+    INFO = 'INFO',
+    PARAM_ERROR = 'PARAM_ERROR',
+    ERROR = "ERROR"
+}
+
+export enum AzureResourceProviderType {
+    QnA = 'Microsoft.CognitiveServices',
+    Luis = 'Microsoft.CognitiveServices',
+    CosmosDB = 'Microsoft.DocumentDB',
+    BlobStorage = 'Microsoft.Storage',
+    ApplicationInsights = 'Microsoft.Insights',
+    WebApp = 'Microsoft.Web',
+    Bot = 'Microsoft.BotService',
+}
+
+export class AzureProvisionUtil {
+
+    /**
+     * credentials from azure login
+     */
+    private credentials: any;
+
+    /**
+     * customized injected logger
+     */
+    private logger: any;
+
+    constructor(credentials: any) {
+        this.credentials = credentials;
+    }
+
+    /**
+     * GetUserSubscriptions based on the credentials of users
+     */
+    public async GetUserSubscriptions(): Promise<Array<Subscription>> {
+        try {
+            const subscriptionClient = new SubscriptionClient(this.credentials);
+            const subscriptionsResult = await subscriptionClient.subscriptions.list();
+            if (subscriptionsResult._response.status >= 300) {
+                this.logger({
+                    status: AzureAPIStatus.ERROR,
+                    message: subscriptionsResult._response.bodyAsText
+                });
+                return []
+            }
+            return subscriptionsResult._response.parsedBody;
+        }
+        catch (err) {
+            this.logger({
+                status: AzureAPIStatus.ERROR,
+                message: JSON.stringify(err, Object.getOwnPropertyNames(err)),
+            });
+            return []
+        }
+    }
+
+    /**
+     * Get supported regions for differernt resource type
+     */
+    public async GetSupportedRegionsByType(subscriptionId: string, resourceType: AzureResourceProviderType): Promise<Array<string>> {
```

Comment:
What is this going to be used for?

Reply:
> What is this going to be used for?

I think it's used for selecting appropriate region for different resources